### Contents

01 Import libraries and dataset

02 Mean of order_number, grouped by department

03 Loyalty flag

04 Spending habits by loyalty category

05 Low/high spender categories

06 Order frequency flag

07 Export dataframe

01 Import libraries and dataset

In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
path = r"C:\Users\cathe\OneDrive\Data Analysis\2 4 Instacart Basket Analysis\02 Data"

In [4]:
df = pd.read_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge_200225.pkl'))

In [5]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

     price_range_loc     busiest_day busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular day        Average orders   
1  Mid-range product  Regularly busy  Regular day        Average orders   
2  Low-range product  Regularly busy  Regular day        Average orders   
3  Low-range product  Regularly busy  Regular day        Average orders   
4  Low-range product  Regularly busy  Regular day        Average orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer

02 Mean of order_number, grouped by department

In [7]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.666130
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

There are minor differences from the results obtained from the subset, but the differences are not huge.  The sample used previously was not chosen randomly: it picked customers with lower user_ids, ie those who registered with Instacart earliest, but the sample nevertheless seems to be representative of the population.

I don't think mean order_number gives the information that the exercise implies.  If two customers both ordered from department 5 once, one in their first order and one in their 15th order, the second one would be a more significant contribution to mean order_number for department 5, although both have shopped in that department once.

03 Loyalty flag

The head of the dataframe above shows the loyalty flag column already present, so I will reproduce the code used below but I won't run it again.

In [ ]:
# Create max_order column

df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [ ]:
# Create loyalty_flag column (stage 1)

df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal  customer'

In [ ]:
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [ ]:
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [15]:
df['loyalty_flag'].value_counts()

loyalty_flag
Regular customer    15876363
Loyal customer      10284010
New customer         6243788
Name: count, dtype: int64

04 Spending habits by loyalty category

In [17]:
df.groupby('loyalty_flag').agg({'prices': ['min', 'max', 'mean']})

prices                
                    min   max      mean
loyalty_flag                           
Loyal customer      1.0  25.0  7.774439
New customer        1.0  25.0  7.802283
Regular customer    1.0  25.0  7.799262

There is very little difference between the spending habits of loyal, regular and new customers.  Each category has the same minimum and maximum spend (`$1` and `$25` respectively).  The average price paid per item differs very little: `$7.80` per item for new and regular customers, and 3 cents less for loyal customers (`$7.77` per item).

05 Low/high spender categories

In [19]:
# Create new column av_item_spend

df['av_item_spend'] = df.groupby('user_id')['prices'].transform(np.mean)

C:\Users\cathe\AppData\Local\Temp\ipykernel_11488\1028451295.py:3: FutureWarning: The provided callable <function mean at 0x000001BD3512CEA0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df['av_item_spend'] = df.groupby('user_id')['prices'].transform(np.mean)


In [21]:
# Check result

df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

     price_range_loc     busiest_day busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular day        Average orders   
1  Mid-range product  Regularly busy  Regular day        Average orders   
2  Low-range product  Regularly busy  Regular day        Average orders   
3  Low-range product  Regularly busy  Regular day        Average orders   
4  Low-range product  Regularly busy  Regular day        Average orders   

   max_order  loyalty_flag  av_item_spend  
0         10  New customer       6.367797  
1         10  New customer       6.367797  
2         10  New customer       6.367797  
3         10  New customer       6.367797  
4         10  New customer       6.367797

This looks right - first one I tried had different numbers for the same user_id so was clearly wrong!

In [23]:
# Create spend_level flag (stage 1)

df.loc[df['av_item_spend'] >= 10, 'spend_level'] = 'High spender'

In [25]:
df.loc[df['av_item_spend'] < 10, 'spend_level'] = 'Low spender'

In [27]:
# Check result

df['spend_level'].value_counts(dropna = False)

spend_level
Low spender     32284201
High spender      119960
Name: count, dtype: int64

06 Order frequency flag

In [29]:
# Create median_days_since_prior_order column

df['median_days_since_prior_order'] = df.groupby('user_id')['days_since_prior_order'].transform(np.median)

C:\Users\cathe\AppData\Local\Temp\ipykernel_11488\2368184807.py:3: FutureWarning: The provided callable <function median at 0x000001BD3524FB00> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df['median_days_since_prior_order'] = df.groupby('user_id')['days_since_prior_order'].transform(np.median)


In [31]:
# Check result

df[['user_id', 'days_since_prior_order', 'median_days_since_prior_order']].head(60)

user_id  days_since_prior_order  median_days_since_prior_order
0         1                     NaN                           20.5
1         1                     NaN                           20.5
2         1                     NaN                           20.5
3         1                     NaN                           20.5
4         1                     NaN                           20.5
5         1                    15.0                           20.5
6         1                    15.0                           20.5
7         1                    15.0                           20.5
8         1                    15.0                           20.5
9         1                    15.0                           20.5
10        1                    15.0                           20.5
11        1                    21.0                           20.5
12        1                    21.0                           20.5
13        1                    21.0                           20.5
14        1                    21.0                           20.5
15        1                    21.0                           20.5
16        1                    29.0                           20.5
17        1                    29.0                           20.5
18        1                    29.0                           20.5
19        1                    29.0                           20.5
20        1                    29.0                           20.5
21        1                    28.0                           20.5
22        1                    28.0                           20.5
23        1                    28.0                           20.5
24        1                    28.0                           20.5
25        1                    28.0                           20.5
26        1                    28.0                           20.5
27        1                    28.0                           20.5
28        1                    28.0                           20.5
29        1                    19.0                           20.5
30        1                    19.0                           20.5
31        1                    19.0                           20.5
32        1                    19.0                           20.5
33        1                    20.0                           20.5
34        1                    20.0                           20.5
35        1                    20.0                           20.5
36        1                    20.0                           20.5
37        1                    20.0                           20.5
38        1                    14.0                           20.5
39        1                    14.0                           20.5
40        1                    14.0                           20.5
41        1                    14.0                           20.5
42        1                    14.0                           20.5
43        1                    14.0                           20.5
44        1                     0.0                           20.5
45        1                     0.0                           20.5
46        1                     0.0                           20.5
47        1                     0.0                           20.5
48        1                     0.0                           20.5
49        1                     0.0                           20.5
50        1                    30.0                           20.5
51        1                    30.0                           20.5
52        1                    30.0                           20.5
53        1                    30.0                           20.5
54        1                    30.0                           20.5
55        1                    30.0                           20.5
56        1                    30.0                           20.5
57        1                    30.0                           20.5
58        1                    30.0                

In [33]:
# Create user-defined function

def order_frequency(row):

    if row['median_days_since_prior_order'] <= 10:
        return 'Frequent customer'
    elif (row['median_days_since_prior_order'] > 10) and (row['median_days_since_prior_order'] <= 20):
        return 'Regular customer'
    elif row['median_days_since_prior_order'] > 20:
        return 'Non-frequent customer'
    else: return 'Not enough data'

In [35]:
# Apply to dataframe to create column

df['order_frequency'] = df.apply(order_frequency, axis=1)

In [37]:
# Check output

df['order_frequency'].value_counts()

order_frequency
Frequent customer        21559380
Regular customer          7208433
Non-frequent customer     3636343
Not enough data                 5
Name: count, dtype: int64

07 Export dataframe

In [39]:
df.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge_200225_2.pkl'))